File: app/game/army.js
import { Piece, Side } from "./piece.js";

export class Army {
  /**
   * @type {Piece[]}
   */
  pieces;

  /**
   * @type {Side}
   */
  side;

  /**
   * @param {Side} side
   */
  constructor(side) {
    this.side = side;
    this.pieces = [];

    this.pieces.push(new Piece(10)); // marshal
    this.pieces.push(new Piece(9)); // general
    this.pieces.push(...this.makePieces(8, 2)); // colonel
    this.pieces.push(...this.makePieces(7, 3)); // major
    this.pieces.push(...this.makePieces(6, 4)); // captain
    this.pieces.push(...this.makePieces(5, 4)); // lieutenant
    this.pieces.push(...this.makePieces(4, 4)); // sergeant
    this.pieces.push(...this.makePieces(3, 5)); // miner
    this.pieces.push(...this.makePieces(2, 8)); // scout
    this.pieces.push(new Piece(1)); // spy
    this.pieces.push(...this.makePieces(0, 6)); // bomb
    this.pieces.push(new Piece(-1)); // flag

    // Assign side to each piece
    this.pieces.forEach((piece) => {
      piece.side = this.side;
    });
  }

  /**
   * Creates an array of pieces with a given rank and count
   * @param {number} rank
   * @param {number} count
   * @returns {Piece[]}
   */
  makePieces(rank, count) {
    return Array.from({ length: count }, () => new Piece(rank));
  }

  /**
   * Find any piece by rank
   * @param {number} rank
   * @returns {Piece}
   * @throws {Error} if no piece with the rank is found
   */
  findAnyByRank(rank) {
    const found = this.pieces.find((p) => p.rank === rank);
    if (!found) {
      throw new Error(`No piece with rank ${rank} found`);
    }
    return found;
  }

  /**
   * Shuffle the pieces array in place using Fisher-Yates algorithm
   */
  shuffle() {
    for (let i = this.pieces.length - 1; i > 0; i--) {
      const j = Math.floor(Math.random() * (i + 1));
      [this.pieces[i], this.pieces[j]] = [this.pieces[j], this.pieces[i]];
    }
  }
}

File: src/main/java/game/Army.java
package game;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class Army {

  public List<Piece> pieces;
  Side side;

  public Army(Side side) {
    this.side = side;
    pieces = new ArrayList<>();
    pieces.add(new Piece(10)); // marshal
    pieces.add(new Piece(9)); // general
    pieces.addAll(makePieces(8, 2)); // colonel
    pieces.addAll(makePieces(7, 3)); // major
    pieces.addAll(makePieces(6, 4)); // captain
    pieces.addAll(makePieces(5, 4)); // lieutenant
    pieces.addAll(makePieces(4, 4)); // sergeant
    pieces.addAll(makePieces(3, 5)); // miner
    pieces.addAll(makePieces(2, 8)); // scout
    pieces.add(new Piece(1)); // spy
    pieces.addAll(makePieces(0, 6)); // bomb
    pieces.add(new Piece(-1)); // flag

    // Mark as belonging to army
    pieces.forEach(x -> x.side = this.side);
  }

  private List<Piece> makePieces(int rank, int count) {
    return IntStream.range(0, count).mapToObj(i -> new Piece(rank)).collect(Collectors.toList());
  }

  public Piece findAnyByRank(int x) {
    return this.pieces.stream().filter(p -> p.rank == x).findAny().get();
  }
}
